On GPU, TestTreeRecombination is failing

### What's the problem?

On NVIDIA GPU, TestTreeRecombination::test_reduce is failing.

### How can we reproduce the issue?

Run  tests/unit/test_solvers.py::TestTreeRecombination::test_reduce

### Python version

3.12

### Package version

0.3.1

### Operating system

Ubuntu 24.04 with NVIDIA GPU

### Other packages

jax[cuda12]

### Relevant log output

FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-random-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-random-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-partial-null-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-partial-null-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-partial-null-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-partial-null-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-partial-null-explicit-context2-False] - assert 16 <= Array(11, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-partial-null-explicit-context2-True] - assert 16 <= Array(11, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-random-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-random-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-random-explicit-context2-False] - assert 16 <= Array(11, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-random-explicit-context2-True] - assert 16 <= Array(11, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-null-implicit-explicit-context0-False] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-null-implicit-explicit-context0-True] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-null-explicit-context2-False] - assert 14 <= Array(1, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-null-explicit-context2-True] - assert 14 <= Array(1, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-null-implicit-explicit-context0-False] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-null-implicit-explicit-context0-True] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-full_rank-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-full_rank-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-full_rank-explicit-context2-False] - assert 6 <= Array(4, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-full_rank-explicit-context2-True] - assert 6 <= Array(4, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-full_rank-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-full_rank-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-rank_deficient-implicit-explicit-context0-False] - assert 3 <= (4 - (4 - Array(2, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-rank_deficient-implicit-explicit-context0-True] - assert 3 <= (4 - (4 - Array(2, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-rank_deficient-explicit-context2-False] - assert 3 <= Array(2, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-rank_deficient-explicit-context2-True] - assert 3 <= Array(2, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-rank_deficient-implicit-explicit-context0-False] - assert 3 <= (4 - (4 - Array(2, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-rank_deficient-implicit-explicit-context0-True] - assert 3 <= (4 - (4 - Array(2, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-implicit-explicit-context0-False] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-implicit-explicit-context0-True] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-implicit-context1-False] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-implicit-context1-True] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-explicit-context2-False] - assert 84 <= Array(1, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-explicit-context2-True] - assert 84 <= Array(1, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-excessive_test_functions-implicit-explicit-context0-False] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-excessive_test_functions-implicit-explicit-context0-True] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-excessive_test_functions-implicit-context1-False] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-excessive_test_functions-implicit-context1-True] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,7 +19,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - `MMD.compute` no longer returns `nan`. (https://github.com/gchq/coreax/issues/855)
 - CaratheodoryRecombination now generates correct coresets on GPU machines
-  (https://github.com/gchq/coreax/issues/853)
+  (https://github.com/gchq/coreax/pull/874)
 
 ### Changed
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,6 +18,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - `MMD.compute` no longer returns `nan`. (https://github.com/gchq/coreax/issues/855)
+- CaratheodoryRecombination now generates correct coresets on GPU machines
+  (https://github.com/gchq/coreax/issues/853)
 
 ### Changed
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,8 +18,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - `MMD.compute` no longer returns `nan`. (https://github.com/gchq/coreax/issues/855)
-- CaratheodoryRecombination now generates correct coresets on GPU machines
-  (https://github.com/gchq/coreax/pull/874)
+- Corrected an implementation error in `coreax.solvers.CaratheodoryRecombination`
+  which caused numerical instability when using either `CaratheodoryRecombination`
+  or `TreeRecombination` on GPU machines. (https://github.com/gchq/coreax/pull/874, see
+  also https://github.com/gchq/coreax/issues/852 and
+  https://github.com/gchq/coreax/issues/853)
 
 ### Changed
 

--- a/coreax/solvers/recombination.py
+++ b/coreax/solvers/recombination.py
@@ -248,7 +248,7 @@ class CaratheodoryRecombination(RecombinationSolver[Data, None]):
                 basis_vector > 0, _weights / basis_vector, jnp.inf
             )
             elimination_index = jnp.argmin(elimination_condition)
-            eliminated = eliminated.at[elimination_index].set(1)
+            eliminated = eliminated.at[elimination_index].set(True)
             elimination_rescaling_factor = elimination_condition[elimination_index]
             # Equation 4: Eliminate the selected weight and redistribute its mass.
             # NOTE: Equation 5 is implicit from Equation 4 and is performed outside

--- a/coreax/solvers/recombination.py
+++ b/coreax/solvers/recombination.py
@@ -141,7 +141,6 @@ class _EliminationState(NamedTuple):
     weights: Shaped[Array, " n"]
     nodes: Shaped[Array, "n m"]
     iteration: int
-    eliminated: Shaped[Array, " n"]
 
 
 class CaratheodoryRecombination(RecombinationSolver[Data, None]):
@@ -193,10 +192,6 @@ class CaratheodoryRecombination(RecombinationSolver[Data, None]):
             safe_push_forward_nodes, self.rcond
         )
 
-        # Array to keep track of which nodes have been eliminated (not described by
-        # the base algorithm, but needed to combat numerical errors on GPU systems)
-        eliminated_array = jnp.zeros_like(weights, dtype=bool)
-
         def _eliminate_cond(state: _EliminationState) -> Bool[Array, ""]:
             """
             If to continue the iterative Gaussian-Elimination procedure.
@@ -241,14 +236,13 @@ class CaratheodoryRecombination(RecombinationSolver[Data, None]):
             # - `updated_weights` -> \underline\Beta^{(i)}
             # - `null_space_basis_update` -> d_{l+1}^{(i)}\phi_1^{(i-1)}
             # - `updated_null_space_basis` -> \Psi^{(i))
-            _weights, null_space_basis, basis_index, eliminated = state
+            _weights, null_space_basis, basis_index = state
             basis_vector = null_space_basis[basis_index]
             # Equation 3: Select the weight to eliminate.
             elimination_condition = jnp.where(
                 basis_vector > 0, _weights / basis_vector, jnp.inf
             )
             elimination_index = jnp.argmin(elimination_condition)
-            eliminated = eliminated.at[elimination_index].set(True)
             elimination_rescaling_factor = elimination_condition[elimination_index]
             # Equation 4: Eliminate the selected weight and redistribute its mass.
             # NOTE: Equation 5 is implicit from Equation 4 and is performed outside
@@ -262,21 +256,14 @@ class CaratheodoryRecombination(RecombinationSolver[Data, None]):
                 axes=0,
             )
             updated_null_space_basis = null_space_basis - null_space_basis_update
-            updated_null_space_basis = updated_null_space_basis.at[basis_index].set(0)
-            # Ensure eliminated rows *stay* eliminated.
-            # On GPU systems, small numerical errors can cause previously eliminated
-            # rows to have small but positive values in the basis vector. This line
-            # ensures that this does not happen.
-            updated_null_space_basis = jnp.where(
-                jnp.expand_dims(eliminated, 0), 0, updated_null_space_basis
-            )
+            updated_null_space_basis = updated_null_space_basis.at[
+                :, elimination_index
+            ].set(0)
             return _EliminationState(
-                updated_weights, updated_null_space_basis, basis_index + 1, eliminated
+                updated_weights, updated_null_space_basis, basis_index + 1
             )
 
-        in_state = _EliminationState(
-            safe_weights, largest_null_space_basis, 0, eliminated_array
-        )
+        in_state = _EliminationState(safe_weights, largest_null_space_basis, 0)
         out_weights, *_ = jax.lax.while_loop(_eliminate_cond, _eliminate, in_state)
         coresubset_nodes = _coresubset_nodes(
             safe_push_forward_nodes,

--- a/coreax/solvers/recombination.py
+++ b/coreax/solvers/recombination.py
@@ -141,6 +141,7 @@ class _EliminationState(NamedTuple):
     weights: Shaped[Array, " n"]
     nodes: Shaped[Array, "n m"]
     iteration: int
+    eliminated: Shaped[Array, " n"]
 
 
 class CaratheodoryRecombination(RecombinationSolver[Data, None]):
@@ -192,6 +193,10 @@ class CaratheodoryRecombination(RecombinationSolver[Data, None]):
             safe_push_forward_nodes, self.rcond
         )
 
+        # Array to keep track of which nodes have been eliminated (not described by
+        # the base algorithm, but needed to combat numerical errors on GPU systems)
+        eliminated_array = jnp.zeros_like(weights, dtype=bool)
+
         def _eliminate_cond(state: _EliminationState) -> Bool[Array, ""]:
             """
             If to continue the iterative Gaussian-Elimination procedure.
@@ -210,8 +215,7 @@ class CaratheodoryRecombination(RecombinationSolver[Data, None]):
             :param state: Elimination state information
             :return: Boolean indicating if to continue/exit the elimination loop.
             """
-            *_, basis_index = state
-            return basis_index < null_space_rank
+            return state.iteration < null_space_rank
 
         def _eliminate(state: _EliminationState) -> _EliminationState:
             """
@@ -237,13 +241,14 @@ class CaratheodoryRecombination(RecombinationSolver[Data, None]):
             # - `updated_weights` -> \underline\Beta^{(i)}
             # - `null_space_basis_update` -> d_{l+1}^{(i)}\phi_1^{(i-1)}
             # - `updated_null_space_basis` -> \Psi^{(i))
-            _weights, null_space_basis, basis_index = state
+            _weights, null_space_basis, basis_index, eliminated = state
             basis_vector = null_space_basis[basis_index]
             # Equation 3: Select the weight to eliminate.
             elimination_condition = jnp.where(
                 basis_vector > 0, _weights / basis_vector, jnp.inf
             )
             elimination_index = jnp.argmin(elimination_condition)
+            eliminated = eliminated.at[elimination_index].set(1)
             elimination_rescaling_factor = elimination_condition[elimination_index]
             # Equation 4: Eliminate the selected weight and redistribute its mass.
             # NOTE: Equation 5 is implicit from Equation 4 and is performed outside
@@ -258,13 +263,20 @@ class CaratheodoryRecombination(RecombinationSolver[Data, None]):
             )
             updated_null_space_basis = null_space_basis - null_space_basis_update
             updated_null_space_basis = updated_null_space_basis.at[basis_index].set(0)
+            # Ensure eliminated rows *stay* eliminated.
+            # On GPU systems, small numerical errors can cause previously eliminated
+            # rows to have small but positive values in the basis vector. This line
+            # ensures that this does not happen.
+            updated_null_space_basis = jnp.where(
+                jnp.expand_dims(eliminated, 0), 0, updated_null_space_basis
+            )
             return _EliminationState(
-                updated_weights,
-                updated_null_space_basis,
-                basis_index + 1,
+                updated_weights, updated_null_space_basis, basis_index + 1, eliminated
             )
 
-        in_state = _EliminationState(safe_weights, largest_null_space_basis, 0)
+        in_state = _EliminationState(
+            safe_weights, largest_null_space_basis, 0, eliminated_array
+        )
         out_weights, *_ = jax.lax.while_loop(_eliminate_cond, _eliminate, in_state)
         coresubset_nodes = _coresubset_nodes(
             safe_push_forward_nodes,

